src/space_efficient_shuffler.rs

```diff
@@ -0,0 +1,263 @@
+    /// exactly one of its children is `None`,
+    /// and `descendant_idx` points us towards the next interesting descendant.
+    One {
+        /// At times there might be path in the tree that forms a linked list,
+        /// in the sense that except for the tail they all have exactly one child that's not `None`,
+        /// i.e. comprised only of consecutive `One` except the tail being `Two` or `Leaf`,
+        /// in which case the `descendant_idx` shall only point to indices within the list.
+        ///
+        /// I.e. if there's some path comprised of consecutive `One`, then a `Two`,
+        /// followed by more `One`,
+        /// those `One` in the previous part shall not point to any `One` in the latter part.
+        descendant_idx: NonZeroUsize,
+    },
+    /// This node has two children.
+    Two {
+        sum: usize,
+    },
+    Leaf(&'u mut U),
+}
+
+impl<'u, U: for<'hrtb> User<'hrtb>> SpaceEfficientShuffler<'u, U> {
+    const ERR_DATA_INCONSISTENT: &'static str = "Data strucutre inconsistent";
+
+    fn left(u: usize) -> NonZeroUsize {
+        NonZeroUsize::new(u * 2 + 1).unwrap()
+    }
+    fn right(u: usize) -> NonZeroUsize {
+        NonZeroUsize::new(u * 2 + 2).unwrap()
+    }
+    fn parent(u: usize) -> Option<usize> {
+        u.checked_sub(1).map(|i| i / 2)
+    }
+    fn sibling(u: NonZeroUsize) -> NonZeroUsize {
+        let u = u.get();
+        NonZeroUsize::new(u - 1 + (u % 2) * 2).unwrap()
+    }
+
+    /// Make `SpaceEfficientShuffler::binary_tree` a _complete binary tree_,
+    /// in which all internal nodes are `BinaryTreeNode::Two`,
+    /// and all leaf nodes are `BinaryTreeNode::Leaf`
+    pub(crate) fn new(iter: impl IntoIterator<Item = &'u mut U>) -> Self {
+        let iter = iter.into_iter();
+        let mut binary_tree = Vec::with_capacity(iter.size_hint().0 * 2);
+        binary_tree.extend(iter.map(BinaryTreeNode::Leaf));
+        if binary_tree.is_empty() {
+            // early return s.t. later we may assume non-zero user count
+            return Self { binary_tree };
+        } else {
+            // to make a complete binary tree in which leaf node iff `BinaryTreeNode::Leaf`,
+            // # of internal nodes is exactly one less than (# of leafs i.e. # of users)
+            let mut internal_nodes =
+                Vec::from_iter(std::iter::repeat_n((), binary_tree.len() - 1).map(|_| BinaryTreeNode::None));
+            binary_tree = {
+                internal_nodes.append(&mut binary_tree);
+                internal_nodes
+            };
+        }
+
+        /// Given internal nodes be `BinaryTreeNode::None` and leaf nodes `BinaryTreeNode::Leaf`,
+        /// make every internal node `BinaryTreeNode::Two`:
+        /// internal nodes in a complete binary tree all have two children
+        ///
+        /// Input should be valid indices only.
+        fn init_at<U: for<'hrtb> User<'hrtb>>(i: usize, v: &mut [BinaryTreeNode<U>]) {
```

Comment:
This lives here for calling it outside of `new` is meaningless: it assumes the whole tree is comprised of only `BinaryTreeNode::None` and `BinaryTreeNode::Leaf`, forms a complete binary tree, and leaves are exactly those `BinaryTreeNode::Leaf`.